gh-115999: Add free-threaded specialization for FOR_ITER

Yhg1s opened the pull request:
Add free-threaded versions of existing specialization for FOR_ITER (list, tuples, fast range iterators and generators), without significantly affecting their thread-safety. (Iterating over shared lists/tuples/ranges should be fine like before. Reusing iterators between threads is not fine, like before. Sharing generators between threads is a recipe for significant crashes, like before.)

<!-- gh-issue-number: gh-115999 -->
* Issue: gh-115999
<!-- /gh-issue-number -->

Yhg1s committed:
Add free-threaded specialization for lists and tuples.

Yhg1s committed:
Add specialization for range iterators and generators, both about as
thread-safe as without spcialization (i.e. not much to none at all).

Yhg1s committed:
Add missing ifdef guard.

Yhg1s committed:
Fix copy-paste mistake.

Yhg1s committed:
Regen cases.

colesbury commented on Objects/tupleobject.c:
```diff
@@ -1020,13 +1020,15 @@ tupleiter_next(PyObject *self)
 
-    it->it_seq = NULL;
+    FT_ATOMIC_STORE_SSIZE_RELAXED(it->it_index, -1);
+    FT_ATOMIC_STORE_PTR_RELAXED(it->it_seq, NULL);
```
We should not set `it_seq` to `NULL` in the free threaded build. It's not safe with concurrent calls to `tupleiter_next`.

Yhg1s replied:
Well, I have PR #128637 to fix that, and I want to get that in first. (That's why this is still a draft PR really.)

colesbury commented on Python/bytecodes.c:
```diff
@@ -3015,10 +3015,20 @@ dummy_func(
+            EXIT_IF(Py_TYPE(iter_o) != &PyListIter_Type);
+#ifdef Py_GIL_DISABLED
+            _PyListIterObject *it = (_PyListIterObject *)iter_o;
+            EXIT_IF(it->it_seq == NULL ||
```
`it->it_seq` must never be `NULL` in the free threading build.

colesbury commented on Python/bytecodes.c:
```diff
@@ -3015,10 +3015,20 @@ dummy_func(
         replaced op(_ITER_JUMP_LIST, (iter -- iter)) {
+// For free-threaded Python, the loop exit can happen at any point during item
+// retrieval, so separate ops don't make much sense.
+#ifndef Py_GIL_DISABLED
```
Remove the nested `#ifndef Py_GIL_DISABLED` below (line 3039) if this whole uop is guarded by an `ifndef Py_GIL_DISABLED`.

colesbury commented on Python/bytecodes.c:
```diff
@@ -3015,10 +3015,20 @@ dummy_func(
+            _PyListIterObject *it = (_PyListIterObject *)iter_o;
+            EXIT_IF(it->it_seq == NULL ||
+                    !_Py_IsOwnedByCurrentThread((PyObject *)it->it_seq) ||
+                    !_PyObject_GC_IS_SHARED(it->it_seq));
```
Ideally, we'd check these once on entry to the loop, not on every iteration, but that might require broader changes.

Yhg1s replied:
Yeah, I intentionally left that as a potential future avenue for improvement. It's not how the specializing interpreter is set up, and we have to be pretty damned sure the iterators can't legitimately escape the loop. (It's actually pretty easy to make the interpreter do this with generators, as is, by using a weakref, and for list/tuple/range iterators I'm not certain it can't be done with instrumentation or stack inspection... So I'm not comfortable making those assumptions right now.)

Yhg1s replied:
What I meant to say is that currently the specializing interpreter will not unspecialize the code when the guards fail, it will just unspecialize the current iteration (by jumping to the unspecialized code), and only checking on loop entry means we'd have to unspecialize the code.

Yhg1s committed:
Rework FOR_ITER specialization in the free-threaded build to only apply to
uniquely referenced iterators. This handles the common case of 'for item in
seq' (where 'seq' is a list, tuple or range object) and 'for item in
generator_function()', but not, for example, 'g = gen(...); for item in g:'.

Yhg1s committed:
Fix whitespace for linter.

Yhg1s committed:
Incorporate changes from #128445 into the free-threaded branch of
FOR_ITER_LIST specialization.

Yhg1s committed:
Drop redundant assert.

colesbury commented on Include/internal/pycore_list.h:
```diff
@@ -13,6 +13,10 @@ extern void _PyList_DebugMallocStats(FILE *out);
 extern PyObject* _PyList_GetItemRef(PyListObject *, Py_ssize_t i);
+#ifdef Py_GIL_DISABLED
+// Returns -1 in case of races with other threads.
+extern int _PyList_GetItemRefNoLock(PyListObject *, Py_ssize_t, PyObject **);
```
I think this should output a `_PyStackRef *` instead of a `PyObject **`. The body should probably also be either in a header (or just inlined directly into bytecodes.c), so it's available to be inlined outside of LTO builds.

Yhg1s replied:
Unfortunately inlining it isn't straight-forward, because it requires making list_capacity available for inlining, which in turn requires making the _PyListArray available for inlining. It may still be worth it, but I think it's better to leave it as a follow-up.

colesbury replied:
Ok, that's fine with me.

colesbury commented on Tools/cases_generator/analyzer.py:
```diff
@@ -594,6 +594,7 @@ def has_error_without_pop(op: parser.InstDef) -> bool:
     "_PyInterpreterState_GET",
     "_PyList_AppendTakeRef",
     "_PyList_FromStackRefSteal",
+    "_PyList_GetItemRefNoLock",
```
The `Py_TryXGetRef` used inside `_PyList_GetItemRefNoLock` can escape if there's a race on reading the item.

Yhg1s replied:
D'oh, that's subtle. Thanks, fixed.

mpage commented on Python/bytecodes.c:
```diff
@@ -3068,20 +3101,31 @@ dummy_func(
-            _PyTupleIterObject *it = (_PyTupleIterObject *)iter_o;
             assert(Py_TYPE(iter_o) == &PyTupleIter_Type);
+// For free-threaded Python, the loop exit can happen at any point during item
+// retrieval, so separate ops don't make much sense.
```
It doesn't seem like this comment applies to the tuple specialization. Can you delete it if not?

Yhg1s replied:
Done.

colesbury commented on Python/bytecodes.c:
```diff
@@ -3015,31 +3015,43 @@ dummy_func(
             STAT_INC(FOR_ITER, hit);
             PyListObject *seq = it->it_seq;
             if (seq == NULL || (size_t)it->it_index >= (size_t)PyList_GET_SIZE(seq)) {
                 it->it_index = -1;
```
Why are the `#ifndef Py_GIL_DISABLED` removed?

Yhg1s replied:
You asked me to :) (https://github.com/python/cpython/pull/128798#discussion_r1915286452). This whole block is in the 'else' part of the ifdef on line 3033.

colesbury replied:
😵‍💫

colesbury commented on Python/bytecodes.c:
```diff
@@ -3057,8 +3070,28 @@ dummy_func(
+            int result = _PyList_GetItemRefNoLock(seq, it->it_index, &item);
+            // A negative result means we lost a race with another thread
+            // and we need to take the slow path.
+            EXIT_IF(result < 0);
```
I think this should be `DEOPT_IF` here instead of `EXIT_IF` because `_ITER_NEXT_LIST` is not a guard, but I'm not sure.

Yhg1s replied:
I'm not sure either, because there are a few other places where EXIT_IF is used outside of guards (like the _CALL_BUILTIN_O and _CALL_METHOD_DESCRIPTOR_O ops) and it actually expands to the exact same thing in all of the (current) generated output, but you're probably right. Updated.

mpage commented on Python/bytecodes.c:
```diff
@@ -3015,31 +3015,43 @@ dummy_func(
-            _PyListIterObject *it = (_PyListIterObject *)iter_o;
             assert(Py_TYPE(iter_o) == &PyListIter_Type);
+// For free-threaded Python, the loop exit can happen at any point during item
+// retrieval, so separate ops don't make much sense.
```
Can you expand a little bit on why this doesn't make sense? It'd be nice to keep the structure of the ops the same between the two builds.

Yhg1s replied:
Done.

mpage commented on Python/bytecodes.c:
```diff
@@ -3068,20 +3101,31 @@ dummy_func(
             assert(Py_TYPE(iter_o) == &PyTupleIter_Type);
+// For free-threaded Python, the loop exit can happen at any point during item
+// retrieval, so separate ops don't make much sense.
+#ifdef Py_GIL_DISABLED
+            assert(_PyObject_IsUniquelyReferenced(iter_o));
+#endif
```
This appears a number of times. Maybe hoist it into a macro?

Yhg1s replied:
I don't think it's worth it. Depending on what goes into the macro (just the assert or the ifdef as well?) I think it'd be too easy to miss what it actually meant, or it'd need a name like 'assert_uniquely_referenced_if_no_gil' which just reads clunky.

colesbury commented on Python/bytecodes.c:
```diff
@@ -3015,31 +3015,43 @@ dummy_func(
+// For free-threaded Python, the loop exit can happen at any point during item
+// retrieval, so separate ops don't make much sense.
+#ifdef Py_GIL_DISABLED
+            assert(_PyObject_IsUniquelyReferenced(iter_o));
```
There's an unused variable warning on `iter_o` in optimized `Py_GIL_DISABLED` builds.

Yhg1s replied:
Fixed.

colesbury commented on Python/bytecodes.c:
```diff
@@ -3158,6 +3217,9 @@ dummy_func(
             PyGenObject *gen = (PyGenObject *)PyStackRef_AsPyObjectBorrow(iter);
             DEOPT_IF(Py_TYPE(gen) != &PyGen_Type);
+#ifdef Py_GIL_DISABLED
+            DEOPT_IF(!_PyObject_IsUniquelyReferenced((PyObject *)gen));
```
I don't think we should deopt here. If we don't have a the only reference to `gen`, then we should use atomic compare-exchange to set `gi_frame_state` to `FRAME_EXECUTING`.

(We can do this in a follow-up PR)

Yhg1s replied:
Generators are not thread-safe at all yet (trying to use them from multiple threads at the same time gets you fun crashes), so I haven't put much thought in the specialisation either, and expect we'd tackle this when we tackle thread-safety in generators and frames in general. I added a comment, and I'll file an issue for generator thread-safety if we don't already have one.

colesbury commented on Python/specialize.c:
```diff
@@ -2647,45 +2647,56 @@ int
+#ifdef Py_GIL_DISABLED
+    // Only specialize for uniquely referenced iterators, so that we know
+    // they're only referenced by this one thread. This is more limiting
+    // than we need (event `it = iter(mylist); for item in it:` won't get
```
event -> even ?

Yhg1s replied:
Fixed.

colesbury commented on Python/specialize.c:
```diff
@@ -2647,45 +2647,56 @@ int
+    // specialized) but we don't have a way to check whether we're the only
+    // _thread_ who has access to the object.
+    if (!_PyObject_IsUniquelyReferenced(iter_o))
+        goto failure;
```
Might be worth adding a stat for this failure case.

Yhg1s replied:
Yeah, I've been thinking the same thing, but I think it's too much work for this PR. I'll do it in a follow-up, I want to run some stats on the benchmarks for free-threaded specialization anyway.

colesbury commented on Lib/test/test_free_threading/test_iteration_deopt.py:
```diff
@@ -0,0 +1,154 @@
+            # _only_ (strongly) referenced by the for loop, so that we can
+            # force our loop to deopt mid-way through.
+            it = get_iter(input)
+            ref = _testlimitedcapi.pylong_fromvoidptr(it)
```
I don't understand this test. It looks like you're doing a bunch of unsafe things with pointers. `_PyObject_IsUniquelyReferenced` implies that there are no weak references to the object. If you create a fake weak reference using unsafe C code... then it'll be unsafe??

I don't think we should care if the specialization is thread-safe if people are smuggling objects through `_testlimitedcapi.pylong_asvoidptr`.

colesbury replied:
`_PyObject_IsUniquelyReferenced` ensures that there are no weakrefs because `ob_ref_shared` must be exactly zero including flags. Weak references require that the `ob_ref_shared` flags are at least `_Py_REF_MAYBE_WEAKREF` (the flags can't be zero).

Yhg1s replied:
As I mentioned, the test was mostly for my own assurance that the deopt paths were correct. The pointer shenanigans are just a shortcut for the object escaping, because I didn't want to spend too much time figuring out the most convenient way for objects to escape (or proving to myself that it wasn't possible.) But it _is_ possible for at least the list, tuple and generator iterators to escape through `gc.get_referrers()`, if nothing else. It doesn't work for range iterators, but it may still be possible for them to escape as well. (I rewrote the test using `gc.get_referrers`, skipping range iterators, and I can commit that if you think the test is worthwhile if it doesn't do patently bad things. I don't mind skipping the test entirely as long as we realise objects can still escape in the middle of the loop.)

colesbury replied:
We should not consider `gc.get_referrers` (or `gc.get_objects`) as part of the thread-safety model. It's an unsafe debugging tool, even with the GIL enabled, and can cause crashes by exposing partly constructed objects and objects that are otherwise not accessibly to Python code.

> ...as long as we realise objects can still escape in the middle of the loop

We should assume that these objects do not escape and, when possible, only check the guards once.

Yhg1s replied:
FWIW, I've dropped the test.

Yhg1s committed:
Don't mark _PyList_GetItemRefNoLock as non-escaping.

Yhg1s committed:
Address reviewer comments, and drop test_iteration_deopt.

markshannon commented on Include/internal/pycore_opcode_metadata.h:
```diff
@@ -2055,7 +2055,7 @@ const struct opcode_metadata _PyOpcode_opcode_metadata[266] = {
     [FOR_ITER] = { true, INSTR_FMT_IBC, HAS_ARG_FLAG | HAS_JUMP_FLAG | HAS_ERROR_FLAG | HAS_ERROR_NO_POP_FLAG | HAS_ESCAPES_FLAG },
     [FOR_ITER_GEN] = { true, INSTR_FMT_IBC, HAS_ARG_FLAG | HAS_DEOPT_FLAG },
-    [FOR_ITER_LIST] = { true, INSTR_FMT_IBC, HAS_ARG_FLAG | HAS_JUMP_FLAG | HAS_EXIT_FLAG },
+    [FOR_ITER_LIST] = { true, INSTR_FMT_IBC, HAS_ARG_FLAG | HAS_JUMP_FLAG | HAS_DEOPT_FLAG | HAS_EXIT_FLAG | HAS_ESCAPES_FLAG },
```
Why does this need the `HAS_ESCAPES_FLAG`? How does it escape?

Yhg1s replied:
The mechanism for fetching an item from a shared list includes calling `_Py_TryIncrefCompare`, which has a path which DECREFs.

markshannon replied:
Which path? I'm confused as why an incref would need to decref.

Yhg1s replied:
When the reference is removed from the array that backs the list between the time it's retrieved and it's returned.

https://github.com/python/cpython/blob/180ee43bde99b8ce4c4f1d5237ab191e26118061/Include/internal/pycore_object.h#L595-L610

Added in https://github.com/python/cpython/pull/114512.

markshannon replied:
I don't see how that works. What happens if `*src` is modified after `op != _Py_atomic_load_ptr(src)` but before the function returns?

Regardless, we want to keep `FOR_ITER_LIST` non-escaping in the default build.

Yhg1s replied:
@colesbury do you remember why lists use `_Py_TryXGetRef` (in list_get_item_ref) or why it matters that the src ptr still refers to the same item? Can the object be invalid in some way that still requires us to DECREF it?

colesbury replied:
There are two hazards that `_Py_TryXGetRef` handles:

1) The object `op` may be deallocated between the initial load and the incref. The try-incref, in cooperation with the allocator, handles this case. It returns zero if `op` has zero refcount.
2) The memory block for `op` may be deallocated *and reallocated* for a different object in between the initial load and incref. In that case the try-incref succeeds, but the subsequent check fails and we need to decref `op`.

It does not matter if `*src` is modified after `op != _Py_atomic_load_ptr(src)`. `op` will still be a valid reference to an object that was the next element in the list at some point during the operation. It's always been possible for the list to be concurrently modified between the execution of `FOR_ITER` and subsequent code that uses the result.

See https://peps.python.org/pep-0703/#optimistically-avoiding-locking

> Regardless, we want to keep FOR_ITER_LIST non-escaping in the default build

Is there any performance reason to do so?

brandtbucher replied:
> > Regardless, we want to keep FOR_ITER_LIST non-escaping in the default build
>
> Is there any performance reason to do so?

Just to follow up: this does make JIT code *slightly* worse for traces covering `for` loops over lists. Since the `_ITER_NEXT_LIST` at the top of the loop is now escaping, this will add a new `_CHECK_VALIDITY` uop to the start of each loop body where none exists currently.

Consider this loop:

```py
def f():
    x = 0
    for i in list(range(10_000)):
        x += i
```

This change increases the number of validity checks per loop from 2 to 3. Before:

```
_MAKE_WARM
_SET_IP
_CHECK_PERIODIC
_CHECK_VALIDITY <------------- OLD
_ITER_CHECK_LIST
_GUARD_NOT_EXHAUSTED_LIST
_ITER_NEXT_LIST
_SET_IP
_STORE_FAST_1
_CHECK_VALIDITY <------------- OLD
_LOAD_FAST_0
_LOAD_FAST_1
_GUARD_BOTH_INT
_BINARY_OP_ADD_INT
_SET_IP
_STORE_FAST_0
_JUMP_TO_TOP
```

After:

```
_MAKE_WARM
_SET_IP
_CHECK_PERIODIC
_CHECK_VALIDITY <------------- OLD
_ITER_CHECK_LIST
_GUARD_NOT_EXHAUSTED_LIST
_ITER_NEXT_LIST
_CHECK_VALIDITY_AND_SET_IP <-- NEW
_STORE_FAST_1
_CHECK_VALIDITY <------------- OLD
_LOAD_FAST_0
_LOAD_FAST_1
_GUARD_BOTH_INT
_BINARY_OP_ADD_INT
_SET_IP
_STORE_FAST_0
_JUMP_TO_TOP
```

markshannon replied:
Just to clarify:
`FOR_ITER_LIST`  is escaping because `_PyList_GetItemRefNoLock` is escaping, and `_PyList_GetItemRefNoLock` is escaping because `_Py_TryIncrefCompareStackRef` is escaping.

Either we need to find an alternative approach or make `_Py_TryIncrefCompareStackRef`  non-escaping.

Yhg1s committed:
Merge branch 'main' into for-iter-spec

Yhg1s committed:
Regenerate cases after merge.

Yhg1s committed:
Merge branch 'main' into for-iter-spec

Yhg1s committed:
Make the free-threaded FOR_ITER work in the tier 2 interpreter/jit.

markshannon commented on Python/bytecodes.c:
```diff
@@ -3075,31 +3075,46 @@ dummy_func(
+            PyObject *iter_o = PyStackRef_AsPyObjectBorrow(iter);
+            EXIT_IF(Py_TYPE(iter_o) != &PyListIter_Type);
+#ifdef Py_GIL_DISABLED
+            EXIT_IF(!_PyObject_IsUniquelyReferenced(iter_o));
```
This seems overly strict. 
Checking that `iter_o` is only referenced from this thread should be sufficient.

Does `_Py_IsOwnedByCurrentThread(op) && _Py_atomic_load_ssize_relaxed(&op->ob_ref_shared) == 0)` do that?

Yhg1s replied:
It might seem overly strict, but we have no way of knowing an object is only referenced from one thread. Checking shared refcounts won't do it, because objects can be referenced from other threads _without directly holding a strong reference_. We can only guarantee that this is the only thread that could potentially use the object if we can account for every reference to the object _and_ guarantee that those references are not usable from other threads. We can currently only guarantee that if we know we have the only reference.

Just to be clear: this is the reference to the _iterator_, not the iterable. So `for item in my_list: ...` is fine. Currently, `it = iter(...); for item in it: ...` can't be specialised, but @mpage's changes to make LOAD_FAST borrow the reference from locals will make that automatically work. Any other references to the iterator will mean we can't specialize it (at least not without more escape analysis of references).

markshannon commented on Python/specialize.c:
```diff
@@ -2826,45 +2826,56 @@ int
+    // than we need (even `it = iter(mylist); for item in it:` won't get
+    // specialized) but we don't have a way to check whether we're the only
+    // _thread_ who has access to the object.
+    if (!_PyObject_IsUniquelyReferenced(iter_o))
```
See my above comment

markshannon reviewed (approved):
It's a shame about `FOR_ITER_LIST` becoming escaping.

We need to make `_Py_TryIncrefCompareStackRef` non-escaping, but not in this PR.
Otherwise we'll end up with everything escaping, as more code uses  `_Py_TryIncrefCompareStackRef`.
